resolver-dns/src/main/java/io/netty/resolver/dns/DefaultDnsCache.java

```diff
@@ -0,0 +1,193 @@
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.TimeUnit;
+
+import static io.netty.util.internal.ObjectUtil.checkPositiveOrZero;
+
+/**
+ * Default implementation of {@link DnsCache}, backed by a {@link ConcurrentMap}
+ */
+public class DefaultDnsCache implements DnsCache {
+
+    private final ConcurrentMap<String, List<DnsCacheEntry>> resolveCache = PlatformDependent.newConcurrentHashMap();
+    private final int minTtl;
+    private final int maxTtl;
+    private final int negativeTtl;
+
+    /**
+     * Create a cache that respects the TTL returned by the DNS server
+     * and doesn't cache negative responses.
+     */
+    public DefaultDnsCache() {
+        this(0, Integer.MAX_VALUE, 0);
+    }
+
+    /**
+     * Create a cache
+     * @param minTtl the minimum TTL
+     * @param maxTtl the maximum TTL
+     * @param negativeTtl the TTL for failed queries
+     */
+    public DefaultDnsCache(int minTtl, int maxTtl, int negativeTtl) {
+        this.minTtl = checkPositiveOrZero(minTtl, "minTtl");
+        this.maxTtl = checkPositiveOrZero(maxTtl, "maxTtl");
+        if (minTtl > maxTtl) {
+            throw new IllegalArgumentException(
+                    "minTtl: " + minTtl + ", maxTtl: " + maxTtl + " (expected: 0 <= minTtl <= maxTtl)");
+        }
+        this.negativeTtl = checkPositiveOrZero(negativeTtl, "negativeTtl");
+    }
+
+    @Override
+    public int minTtl() {
+        return minTtl;
+    }
+
+    @Override
+    public int maxTtl() {
+        return maxTtl;
+    }
+
+    @Override
+    public int negativeTtl() {
+        return negativeTtl;
+    }
+
+    @Override
+    public void clearCache() {
+        for (Iterator<Map.Entry<String, List<DnsCacheEntry>>> i = resolveCache.entrySet().iterator(); i.hasNext();) {
+            final Map.Entry<String, List<DnsCacheEntry>> e = i.next();
+            i.remove();
+            cancelExpiration(e.getValue());
+        }
+    }
+
+    @Override
+    public boolean clearCache(String hostname) {
+        boolean removed = false;
+        for (Iterator<Map.Entry<String, List<DnsCacheEntry>>> i = resolveCache.entrySet().iterator(); i.hasNext();) {
+            final Map.Entry<String, List<DnsCacheEntry>> e = i.next();
+            if (e.getKey().equals(hostname)) {
+                i.remove();
+                cancelExpiration(e.getValue());
+                removed = true;
+            }
+        }
+        return removed;
+    }
+
+    @Override
+    public List<DnsCacheEntry> get(String hostname) {
+        return resolveCache.get(hostname);
+    }
+
+    private List<DnsCacheEntry> cachedEntries(String hostname) {
+        List<DnsCacheEntry> oldEntries = resolveCache.get(hostname);
+        final List<DnsCacheEntry> entries;
+        if (oldEntries == null) {
+            List<DnsCacheEntry> newEntries = new ArrayList<DnsCacheEntry>();
```

Comment:
do we want to use a explicit initialCapacity here ? The default may be a bit too high